src/Service/NuRaftLogSnapshot.h

```diff
@@ -8,6 +8,42 @@ namespace RK
 using nuraft::snapshot;
 using nuraft::ulong;
 
+/// Asynchronously snapshot creating task.
+struct SnapTask
+{
+    ptr<snapshot> s;
+    int64_t next_zxid;
+    int64_t next_session_id;
+    int64_t nodes_count;
+    int64_t ephemeral_nodes_count;
+    int64_t session_count;
+    KeeperStore::SessionAndTimeout session_and_timeout;
+    std::unordered_map<uint64_t, Coordination::ACLs> acl_map;
+    KeeperStore::SessionAndAuth session_and_auth;
+    KeeperStore::BucketNodes buckets_nodes;
+    nuraft::async_result<bool>::handler_type when_done;
+
+    SnapTask(const ptr<snapshot> & s_, KeeperStore & store, nuraft::async_result<bool>::handler_type & when_done_)
+        : s(s_), next_zxid(store.zxid), next_session_id(store.session_id_counter), when_done(when_done_)
+    {
+        session_and_timeout = store.getSessionTimeOut();
+        session_count = session_and_timeout.size();
+        session_and_auth = store.getSessionAuth();
+
+        acl_map = store.acl_map.getMapping();
+
+        buckets_nodes = store.dumpDataTree();
```

Comment:
`store.dumpDataTree()` should return std::shared_ptr<KeeperStore::BucketNodes> to avoid data copy.

Reply:
In C++, when a function returns a temporary variable, the compiler typically attempts Return Value Optimization (RVO). It's no copy here, but std::shared_ptr<KeeperStore::BucketNodes> maybe better.

Reply:
I think it is better to use ptr for extremely large object, it is better to avoid dependency of compiler optimization and it will provide same behavier between debug and release mode.